exonum/src/blockchain/mod.rs

```diff
@@ -305,34 +298,52 @@ impl BlockchainMut {
                 call_errors.put(&location, error);
             }
         }
-        // Get tx & state hash.
+
+        let (patch, block) = self.create_block_header(fork, proposer_id, height, tx_hashes);
+        log::trace!("Executing {:?}", block);
+
+        // Calculate block hash.
+        let block_hash = block.object_hash();
+        // Update height.
+        let fork = Fork::from(patch);
+        let schema = Schema::new(&fork);
+        schema.block_hashes_by_height().push(block_hash);
+        // Save block.
+        schema.blocks().put(&block_hash, block);
+        (block_hash, fork.into_patch())
+    }
+
+    fn create_block_header(
+        &self,
+        fork: Fork,
+        proposer_id: Option<ValidatorId>,
+        height: Height,
+        tx_hashes: &[Hash],
+    ) -> (Patch, Block) {
+        // Get last hash.
+        let last_hash = self.inner.last_hash();
```

Comment:
Nit: this variable can be named `prev_hash`, so the `Block` constructor will be a bit more clear.